Fix LXC container implementations

Nomsplease opened the pull request:
# Proposed Changes
<!--- Provide a general summary of your changes -->

- correct some implementation ideas for lxc supprt
- We should not assume the rc.local is empty and just overwrite it,  we can do this better and insert a managed code block
- We also should not assume we are root on reboot handler, and become has been added to this handler
- Also added resetting the lxc modification to a untouched state inside the container, resetting the lxc config in the host may also be a worthwhile adventure to make sure we get as close to original as possible.

## Checklist

- [x] Tested locally
- [x] Ran `site.yml` playbook
- [x] Ran `reset.yml` playbook
- [X] Did not add any unnecessary changes
- [X] Ran pre-commit install at least once before committing
- [X] 🚀

timothystewart6 commented:
FYI @acdoussan   Look ok?

acdoussan commented:
Wow, someone else already using this, pretty cool!

Not assuming the user is root makes total sense, this LGTM.

My assumption was that the hosts would be dedicated for kubernetes since containers have a very low overhead, so I wasn't too worried about stomping over an already existing rc.local. With that being said I don't see any reason to not be nice and not overwrite stuff that is already there.

Have one comment on the impl, otherwise think this looks good to me. Also nice to add this to the clean-up script. 

Unrelated note: maybe I should update the clean-up to remove the container conf update lines as well?

acdoussan commented on roles/lxc/tasks/main.yml:
```diff
@@ -1,7 +1,14 @@
 ---
+- name: Set rc.local fact
+  set_fact:
+    rc_local: "{{ lookup('template', 'templates/rc.local.j2') }}"
```
any reason we cant just use this directly in line 10 for the content, rather than setting a variable?

If we need to keep this variable, think we should name it something more specific to the role, setting a fact that is just called `rc_local` seems a bit risky if something else needs to play with rc.local for some reason.  Maybe `proxmox_lxc_rc_local` to keep in line with the existing variables?

Applies to the cleanup task as well.

Nomsplease replied:
I thought about this, but did not test it. I will validate if we even need to set the fact or not no my next test run of this.

acdoussan commented on roles/reset/tasks/main.yml:
```diff
@@ -54,3 +54,6 @@
     state: absent
+
+- name: Remove LXC modifications if applicable
+  include_tasks: lxc.yml
```
Shouldn't we only include this task list if `proxmox_lxc_configure` is true?

Nomsplease replied:
I thought about that too, but even if it is not true it still guarantees we didn't leave any trace of this behind. We can just add a when to check for it though if we think we should limit the scope.

acdoussan replied:
ah fair actually, if the block isn't there this should just do nothing yeah? probably fine to always run it

Nomsplease replied:
> ah fair actually, if the block isn't there this should just do nothing yeah? probably fine to always run it

Yeah it will check for this managed block and just make sure it is absent from the file. So if this block was never hit in the deploy, it will never be there after wards. It also will not attempt to create the file on the reset either. 

I was more looking at covering all the bases incase someone used a variable they did not mean to and change it between the reset run.

acdoussan replied:
yeah, gotcha, think this is fine as is, good point

timothystewart6 commented:
> Unrelated note: maybe I should update the clean-up to remove the container conf update lines as well?

That would be awesome and thank you for the help!

Nomsplease commented:
@acdoussan I removed the need for set fact in the deploy task and validated it works exactly as it should.

I also added in resetting the LXC container configs as well, made testing the changes faster so figured may as well knock it out in the same reboot on the cluster teardown.

acdoussan commented on roles/reset/tasks/lxc.yml:
```diff
@@ -0,0 +1,61 @@
+---
+- name: Set rc.local fact
```
should probably mirror the task impl and not use a fact no?

acdoussan commented on roles/reset/tasks/lxc.yml:
```diff
@@ -0,0 +1,61 @@
+    create: false
+    state: absent
+
+- name: check for container files that exist on this host
```
since this is created by a separate role, should we have this in a separate task file?

acdoussan commented on roles/reset/tasks/lxc.yml:
```diff
@@ -0,0 +1,61 @@
+    path: /etc/rc.local
+    content: '{{ rc_local }}'
+    create: false
+    state: absent
```
should we still have the reboot handlers when cleaning up?

acdoussan commented:
sorry, couple more thoughts, but appreciate you adding in cleaning up the proxmox conf files!

Had one other thought, the `rc.local` script has a shebang at the beginning, but this will get wrapped in the ansible block comments, I'm not sure if that matters, but we should probably make sure the block gets added to the top of the file, rather than the end? Will look into if this really matters or not

acdoussan commented:
shebang does have to be the first line, [not clear](https://askubuntu.com/questions/9853/how-can-i-make-etc-rc-local-run-on-startup) if it matters for `rc.local` or not. Maybe we remove the line from the templated `rc.local`, and instead have it as a line in file, with firstmatch as true, insertbefore as a match anything regex, and create as true? Would then not need the create on the block in file, that could just stay as is and remove the file perms.

Nomsplease commented:
> sorry, couple more thoughts, but appreciate you adding in cleaning up the proxmox conf files!
> 
> Had one other thought, the `rc.local` script has a shebang at the beginning, but this will get wrapped in the ansible block comments. I'm not sure if that matters, but we should probably make sure the block gets added to the top of the file, rather than the end? Will look into if this really matters or not.

This is a good catch, yes that technically matter as the bang should be the first line. My first thought would be to break the block up. If rc.local exists this should only add the managed block, but if we are creating a file we can add the shebang.

This should be solvable with lineinfile to look for a shebang statement.

acdoussan commented:
cool cool, let me know if you need any help here making those changes, can always separate things out into separate PRs

Nomsplease commented:
Just noting here that I still have this on my to do list, just haven't been able to check this back out and validate. This is in my list for this week.

timothystewart6 commented:
> Just noting here that I still have this on my to do list, just haven't been able to check this back out and validate. This is in my list for this week.

Sounds good, thank you!

Nomsplease commented:
I think we should be good here. I rewrote a lot of how we handle the rc.local file to not blow away an existing file. I also included and validated the lxc configuration and removal extra settings. All of these changes were tested locally and I can confirm they work. I did however find an issue with the implementation, and thats zfs and the overlayfs. I will open an issue on this an look into that.

Also added a clean up task that if the rc.local file only has 1 line or less after we removed our modifications, Im assuming its just a shebang line and deleting the file. There could be an edge case here where someone did not setup a rc.local file correctly and has a one liner or didn't have a shebang when they made a custom file. I don't have an answer on a good way to handle this and I would rather clean up our changes.

Examples of the rc.local modifications.
Existing rc.local file ends up like this:
`root@K3S-01:~# cat /etc/rc.local
#!/bin/bash

echo "Starting up"
echo "We are doing the stuff here"
# BEGIN ANSIBLE MANAGED BLOCK
# Kubeadm 1.15 needs /dev/kmsg to be there, but it's not in lxc, but we can just use /dev/console instead
# see: https://github.com/kubernetes-sigs/kind/issues/662
if [ ! -e /dev/kmsg ]; then
    ln -s /dev/console /dev/kmsg
fi

# https://medium.com/@kvaps/run-kubernetes-in-lxc-container-f04aa94b6c9c
mount --make-rshared /
# END ANSIBLE MANAGED BLOCK`

An new rc.local file will end up like this:
`root@K3S-02:~# cat /etc/rc.local
#!/bin/sh -e
# BEGIN ANSIBLE MANAGED BLOCK
# Kubeadm 1.15 needs /dev/kmsg to be there, but it's not in lxc, but we can just use /dev/console instead
# see: https://github.com/kubernetes-sigs/kind/issues/662
if [ ! -e /dev/kmsg ]; then
    ln -s /dev/console /dev/kmsg
fi

# https://medium.com/@kvaps/run-kubernetes-in-lxc-container-f04aa94b6c9c
mount --make-rshared /
# END ANSIBLE MANAGED BLOCK`
